lib/dynflow/rails/daemon.rb

```diff
@@ -1,32 +1,61 @@
 require 'fileutils'
+require 'get_process_mem'
+require 'dynflow/watchers/memory_consumption_watcher'
 
 module Dynflow
   class Rails
     class Daemon
+      attr_reader :dynflow_memory_watcher_class, :daemons_class
+
+      # make Daemon dependency injection ready for testing purposes
+      def initialize(
+        dynflow_memory_watcher_class = ::Dynflow::Watchers::MemoryConsumptionWatcher,
+        non_default_daemons_class = nil
+      )
+        @dynflow_memory_watcher_class = dynflow_memory_watcher_class
+        @daemons_class = non_default_daemons_class
+      end
+
+      def daemons_class
+        @daemons_class || ::Daemons
+      end
+
       # Load the Rails environment and initialize the executor in this thread.
-      def run(rails_root = Dir.pwd)
+      def run(rails_root = Dir.pwd, options = {})
         STDOUT.puts('Starting Rails environment')
         rails_env_file = File.expand_path('./config/environment.rb', rails_root)
         unless File.exist?(rails_env_file)
           raise "#{rails_root} doesn't seem to be a Rails root directory"
         end
+
+        STDERR.puts("Starting dynflow with the following options: #{options}")
+
         ::Rails.application.dynflow.executor!
+
+        if !options[:memory_limit].nil? && options[:memory_limit] > 0
```

Comment:
Could you please pull in and adapt the tests from https://github.com/theforeman/foreman-tasks/pull/268/files? Just so we don't lose the tests once this gets in.